### PR TITLE
fix decode function with array types

### PR DESCRIPTION
**Pull request by pthomalla**
fix to  https://github.com/poanetwork/ex_abi/issues/12

### CODE REVIEW

File: lib/abi.ex [hunk 1]
```diff
@@ -89,7 +93,7 @@ defmodule ABI do
   otherwise this won't work as expected. If you are decoding transaction input data
   the identifier is the first four bytes and should already be there.
 
-  To find and decode events instead of functions, see `find_and_decode_event/6`
+    To find and decode events instead of functions, see `find_and_decode_event/6`
```
Comment (pgebal) [inline lib/abi.ex]: Unnecessary space.

File: test/abi/type_decoder_test.exs [hunk 2]
```diff
@@ -24,38 +25,21 @@ defmodule ABI.TypeDecoderTest do
     end
   end
 
-  describe "decode_raw" do
+  describe "decode raw" do
```
Comment (pgebal) [inline test/abi/type_decoder_test.exs]: This does not longer test decode_raw but just decode.

File: lib/abi/type_encoder.ex [hunk 3]
```diff
@@ -277,9 +286,18 @@ defmodule ABI.TypeEncoder do
     n < :math.pow(2, max_bits - 1) * -1 + 1 || n > :math.pow(2, max_bits - 1) - 1
   end
 
+  # TODO change to ExthCrypto.Math.mod when it's fixed ( mod(-75,32) == 21 )
+  def mod(x, n) do
```
Comment (pgebal) [inline lib/abi/type_encoder.ex]: It returns incorrect remainder for negative arguments.

File: lib/abi/type_decoder.ex [hunk 4]
```diff
@@ -146,143 +159,58 @@ defmodule ABI.TypeDecoder do
...
   """
   def decode_raw(binary_data, types) do
-    decode_raw(binary_data, types, 0)
-  end
-
-  defp decode_raw(binary_data, types, initial_cursor_offset) do
-    {reversed_result, _} =
-      Enum.reduce(types, {[], initial_cursor_offset}, fn type, {acc, cursor_offset} ->
-        do_decode_raw(binary_data, type, cursor_offset, acc)
+    {reversed_result, binary_rest} =
```
Comment (pgebal) [inline lib/abi/type_decoder.ex]: That's a breaking change. Please check, if we could preserve the old API.

**Comment by ayrat555** [comment]
@pgebal  thanks for the review!
@pthomalla I'll merge it when @pgebal 's comments are resolved

**Comment by pthomalla** [comment]
```elixir
function_selector =
  ABI.FunctionSelector.parse_specification_item(%{
    "type" => "function",
    "name" => "baz",
    "inputs" => [
      %{"name" => "value1", "type" => "uint32[]"},
      %{"name" => "value1", "type" => "uint32[]"}
    ],
    "outputs" => []
  })

data = ABI.TypeEncoder.encode([[1, 2, 3, 4], [4, 5, 6]], function_selector)
ABI.TypeDecode.decode(data, function_selector)
```
it works but method_id in FunctionSelector shout  be set

**Pull request merged**